File: backend/app/fish/utils/card.py
from typing import List, Dict

from ..models.composite import Card, HalfSuit
from ..models.enums import CardRank, CardSuit, HalfSuits


def create_all_cards() -> List[Card]:
    """Returns a list of all cards in the 54 card deck"""
    cards = []
    for rank in CardRank:
        if rank != CardRank.JOKER and rank != CardRank.CUT:
            for suit in CardSuit:
                if suit != CardSuit.JOKER:
                    cards.append(Card(rank=rank, suit=suit))
        else:
            cards.append(Card(rank=rank, suit=CardSuit.JOKER))
    return cards

def create_half_suits() -> Dict[HalfSuits, HalfSuit]:
    """Returns a dictionary of all half suits"""
    return { hs: HalfSuit(half_suit=hs) for hs in HalfSuits }

File: frontend/src/types/card.ts
import { id_to_rank_suit, unique_card_id } from "@/utils/utils";
import { CardSuit, HalfSuits, CardRank } from "./enums"
import { half_suit_id } from "@/utils/card";

export class Card {
  rank: CardRank;
  suit: CardSuit;
  half_suit_id: HalfSuits;
  id: string;

  constructor(id: string) {
    const res = id_to_rank_suit(id)
    if (!res.success) throw new Error("Invalid Card ID")

    this.rank = res.data!.rank
    this.suit = res.data!.suit
    this.half_suit_id = half_suit_id(this.rank, this.suit)
    this.id = id
  }

  cmp = (other: Card) => {
    if (this.half_suit_id != other.half_suit_id) return 0
    if (this.id == other.id) return 0

    const mapp: Record<CardRank, number> = {
      [CardRank.TWO]: 2,
      [CardRank.THREE]: 3,
      [CardRank.FOUR]: 4,
      [CardRank.FIVE]: 5,
      [CardRank.SIX]: 6,
      [CardRank.SEVEN]: 7,
      [CardRank.EIGHT]: 8,
      [CardRank.NINE]: 9,
      [CardRank.TEN]: 10,
      [CardRank.JACK]: 11,
      [CardRank.QUEEN]: 12,
      [CardRank.KING]: 13,
      [CardRank.ACE]: 14,
      [CardRank.CUT]: 15,
      [CardRank.JOKER]: 16,
    }
    if (mapp[this.rank] == mapp[other.rank]) return this.suit < other.suit ? -1 : 1
    return mapp[this.rank] < mapp[other.rank] ? -1 : 1
  }
}

export function create_all_cards() {
  var cards: Card[] = []
  for (const rank of Object.values(CardRank)) {
    if (rank != CardRank.JOKER && rank != CardRank.CUT) {
      for (const suit of Object.values(CardSuit)) {
        if (suit != CardSuit.JOKER)
          cards.push(new Card(unique_card_id(rank, suit)))
      }
    } else
      cards.push(new Card(unique_card_id(rank, CardSuit.JOKER)))
  }
  return cards
}
